### PR TITLE
Wallet tests and wallet notification tests

### PR DESCRIPTION
**Pull request by Neurosploit**
Should be pretty straight forward tested the wallet model methods and made some improvements there based on the test results.

I would like to suggest as a future improvement we create an hdpath parser that returns an hdpath model to centralize that behavior. It will also make the meaning of the actual data being separated by slashes more obvious to other programmers.

Up next will be the manager and syncmanager. After that I will move to the watchonlywallet.

### CODE REVIEW

File: Stratis.Bitcoin/Wallet/Wallet.cs [hunk 1]
```diff
@@ -296,18 +330,21 @@ public HdAddress GetLastUsedAddress(bool isChange)
         /// <returns></returns>
         public IEnumerable<TransactionData> GetTransactionsById(uint256 id)
         {
-            var addresses = this.ExternalAddresses.Concat(this.InternalAddresses);
-            return addresses.SelectMany(a => a.Transactions.Where(t => t.Id == id));
-        }
+            Guard.NotNull(id, nameof(id));
+
+            IEnumerable<HdAddress> addresses = GetCombinedAddresses();
```
Comment (dangershony) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: this.GetCombinedAddresses()

**Comment by dangershony** [comment]
Yeah there is a BIP44 parser in the [coinvault project](https://github.com/CoinVault/dotblock/blob/master/src/Blockchain.Protocol.Bitcoin/Address/ExtendedKeyPathBip44.cs) we just need to bring it over.

This has been discussed indeed.

File: Stratis.Bitcoin/Wallet/Wallet.cs [hunk 2]
```diff
@@ -387,7 +442,20 @@ public HdAddress()
         /// </returns>
         public bool IsChangeAddress()
         {
-            return int.Parse(this.HdPath.Split('/')[4]) == 1;
+            if (string.IsNullOrWhiteSpace(this.HdPath))
+                return false;
```
Comment (bokobza) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: maybe throw an exception here.
we don't support non-HD addresses so there should always be an address here.
if not, something is wrong and we need to know about it.

**Comment by dangershony** [comment]
We'll push this after the release

File: Stratis.Bitcoin/Wallet/Wallet.cs [hunk 3]
```diff
@@ -387,7 +442,20 @@ public HdAddress()
...
+
+            var hdPathParts = this.HdPath.Split('/');
+            if (hdPathParts.Length < 5)            
+                return false;
```
Comment (bokobza) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: should be an exception here too if it's not formed properly

File: Stratis.Bitcoin/Wallet/Wallet.cs [hunk 4]
```diff
@@ -387,7 +442,20 @@ public HdAddress()
...
+
+            int result = 0;
+            if (int.TryParse(hdPathParts[4], out result))
```
Comment (bokobza) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: in this class you're suggesting we do (100% with you on this), we should maybe have something to validate whether an hdpath is formed correctly. 
so it'd have the checks you're doing here: 
- check number of parts (6)
- check int in part 5
etc...

**Comment by bokobza** [comment]
Thanks @Neurosploit .
I'm not done with the review yet, I'm busy with the release for today, but I'll take care of it next week.

File: Stratis.Bitcoin/Wallet/Wallet.cs (hunk 4 above)
Comment (Neurosploit) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: I agree fully with this validation.

File: Stratis.Bitcoin/Wallet/Wallet.cs [hunk 5]
```diff
@@ -320,8 +357,26 @@ public IEnumerable<TransactionData> GetSpendableTransactions()
...
+            var addresses = this.GetCombinedAddresses();
+            return addresses.Where(t=> t.Transactions != null).Where(a => a.Transactions.Any(predicate));
+        }
+
+        private IEnumerable<HdAddress> GetCombinedAddresses()
+        {
+            IEnumerable<HdAddress> addresses = new List<HdAddress>();
+            if (this.ExternalAddresses != null)
+            {
+                addresses = addresses.Concat(this.ExternalAddresses);
```
Comment (bokobza) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: I think here you can probably just do addresses = this.ExternalAddresses and concat the internal addresses if != null

**Comment by bokobza** [comment]
only one more comment and we're good to go :bicyclist:

File: Stratis.Bitcoin/Wallet/Wallet.cs (hunk 5 above)
Comment (Neurosploit) [inline Stratis.Bitcoin/Wallet/Wallet.cs]: Consider that done.

**Comment by bokobza** [comment]
lovely, thanks @Neurosploit

**Pull request merged**